pulp_2to3_migration/app/migrations/0014_gzip_repodata.py

```diff
@@ -0,0 +1,91 @@
+# Generated by Django 2.2.15 on 2020-09-21 10:12
+
+from django.db import migrations, models
+
+import gzip
+
+
+def gzip_repodata(apps, schema_editor):
+    """
+    Compress repodata and store different types of it in a separate field.
+
+    Also remove uncompressed repodata.
+    """
+
+    def gzip_pkg_repodata(pkg):
+        pkg.primary_template_gz = gzip.zlib.compress(pkg.repodata.get('primary').encode())
+        pkg.other_template_gz = gzip.zlib.compress(pkg.repodata.get('other').encode())
+        pkg.filelists_template_gz = gzip.zlib.compress(pkg.repodata.get('filelists').encode())
```

Comment:
TL;DR: This is expensive and takes a lot of time. 20K RPMs migrate 2-3 mins. The majority of them will never be used. Is it worth migrating only those which will be used?

Pulp2Rpm and Pulp2Srpm currently have repodata in uncompressed form in the repodata field.
Repodata field is only used  to create a Package object in Pulp3. When Pulp2Rpm object is migrated, and created package is linked to it, I believe Pulp2Rpm.repodata will never be used again. Should we just set all *_template_gz to null in such case and that's it? Am I forgetting any corner case when Package can disappear from the Pulp2Rpm relation?

Why is this migration needed? - for the case when user started RPM migration and it didn't fully finish (e.g. OOM happened), then they are upgrading (because they need this memory fix) and  trigger RPM migration again. Those items which were already premigrated but not migrated yet, need this data migration.

Reply:
To avoid a need for such migrations, I think we might need a `reset`  or `force` option. For the case when something went south, user is able to start again. Thoughts?

Reply:
how will you identify which packages will be used?
+1 on unsetting the repodata, git grep did not show any other usage of it

Reply:
what will that reset or force option do?

Reply:
> how will you identify which packages will be used?

`Pulp2Rpm.objects.filter(pulp2content__pulp3_content=None)`

Reply:
> what will that reset or force option do?

Somehow let the migration happen from scratch. E.g. remove all premigrated data. We need to braindump.

Reply:
> To avoid a need for such migrations, I think we might need a `reset` or `force` option. For the case when something went south, user is able to start again. Thoughts?
I am in favour of adding such option, as you mentioned we should meet and brainstorm



> > how will you identify which packages will be used?
> 
> `Pulp2Rpm.objects.filter(pulp2content__pulp3_content=None)`

I think it makes sense to filter only the needed content, this will improve the speed of the migration.
Speaking of whether to remove the 'repodata' field - i would rather let it be in case in the future, for some reason, we might need to get to that info. Yes, some of the content will have empty repodata and other not but I do not see this as big issue. Thoughts?

Reply:
@ipanova , after a discussion with @dralley , I think I'm convinced that we can leave this migration as is. It doesn't affect katello users, since it's only for RPM content and RPM migration is not a part of katello yet. As for pure pulp users, there will be very few of them affected (upgrade in the middle of broken migration) and only if their setup is very large. So we go with consistency and migrate all data.

Reply:
@goosemania I'm not sure if it warrants much more thought, but I'm going to test your PR, and I'm going to try a thread pool.  I've seen some blog posts to suggest that the Gzip module bypasses the global interpreter lock so compressing / decompressing multiple things in true parallel is possible.  But I approve of the current approach regardless.

Reply:
Thanks @dralley , to be honest, I think for the amount of cases we discussed, it's likely not worth the effort.

Reply:
@goosemania On my dev laptop it only took about 30 seconds to run the django migration with 20k RPMs, but I imagine that will be a fair bit faster than the majority of production boxes. I consider it a good sign though.  Approved.